src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Utilities/NullableHelpers/NullableHelpers.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal static class NullableHelpers
+    {
+        /// <summary>
+        /// Given an encapsulating operation, returns true if the symbol passed in
+        /// is ever assigned a possibly null value as determined by nullable flow state. Returns
+        /// null if no references are found, letting the caller determine what to do with that information
+        /// </summary>
+        public static bool? IsSymbolAssignedPossiblyNullValue(SemanticModel semanticModel, IOperation containingOperation, ISymbol symbol)
+        {
+            var references = containingOperation.DescendantsAndSelf()
+                .Where(o => IsSymbolReferencedByOperation(o, symbol));
+
+            var hasReference = false;
+
+            foreach (var reference in references)
+            {
+                hasReference = true;
+
+                if (reference is IForEachLoopOperation forEachLoop)
+                {
+                    var foreachInfo = semanticModel.GetForEachStatementInfo((ForEachStatementSyntax)forEachLoop.Syntax);
```

Comment:
it always worries me to aassume anything about the syntax we get back from an iop.

Reply:
And indeed, this one should. Make sure you have a test for deconstruction foreach, it should fail this cast. `CommonForEachStatementSyntax` is what you're looking for.

Reply:
what is deconstruction foreach? 

Changed to be `CommonForEachStatementSyntax` but my google fu is too weak to find what you're talking about so I can make a test

Reply:
`foreach (var (x, y) ...)`

Reply:
The good news: This is handled by a different code path and not actually checked with `NullableHelpers.IsSymbolAssignedPossiblyNullValue`

The bad news: it deconstructs incorrectly. Nullability seems to get lost entirely

```csharp
#nullable enable

class C
{
    public void Deconstruct(out string? s1, out string? s2)
    {
        s1 = null;
        s2 = null;
    }

    void M(C[] items)
    {
        foreach ([||]var (s1, s2) in items)
        {

        }
    }
}
```

becomes 

```csharp
#nullable enable

class C
{
    public void Deconstruct(out string? s1, out string? s2)
    {
        s1 = null;
        s2 = null;
    }

    void M(C[] items)
    {
        foreach ((string s1, string s2) in items)
        {

        }
    }
}
```